homeassistant/components/weather/__init__.py

```diff
@@ -173,23 +451,69 @@ def ozone(self) -> float | None:
 
     @property
     def visibility(self) -> float | None:
-        """Return the visibility in native units."""
+        """Return the visibility for backward compatibility.
+
+        Should not be set by integrations.
+        """
         return self._attr_visibility
 
     @property
-    def visibility_unit(self) -> str | None:
+    def native_visibility(self) -> float | None:
+        """Return the visibility in native units."""
+        return self._attr_native_visibility
```

Comment:
The deprecated `visibility` reports the native visibility, not the display visibilityso we should return it here:
```suggestion
        if (visibility:= self.visibility) is not None:
            return visibility

        return self._attr_native_visibility
```